---
title: "epic-QTL Data"
author: "Cassandra Buzby"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{epic-QTL Data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

#library(devtools)
#install_github("cbuzby/cybrBSA")
library(cybrBSA)

#Load packages
library(ggplot2)
library(tidyr)
library(reshape2)
library(cowplot)
library(data.table)

library(dplyr)
library(foreach)
library(doParallel)
library(RColorBrewer)
library(circlize)

ggplot2::theme_set(theme_light())
```

```{r, include = FALSE}
cybrConvertParentalAlleles_n <- function(P1 = Wine, P2 = Oak,
                                       P1l = "Wine", P2l = "Oak", Truncate = TRUE, yeast = TRUE){

  P1 %>% mutate(parent = P1l) -> P1
  P2 %>% mutate(parent = P2l) %>% rbind(P1) -> mergeparents

  if(yeast == TRUE){
    ChromKey <- data.frame(chromosomes = c("I", "II", "III", "IV", "V", "VI", "VII", "VIII",
                                           "IX", "X", "XI", "XII", "XIII", "XIV", "XV", "XVI", "M"),
                           CHROM = c("NC_001133.9", "NC_001134.8", "NC_001135.5", "NC_001136.10",
                                     "NC_001137.3", "NC_001138.5", "NC_001139.9", "NC_001140.6",
                                     "NC_001141.2", "NC_001142.9", "NC_001143.9", "NC_001144.5",
                                     "NC_001145.3", "NC_001146.8", "NC_001147.6", "NC_001148.4", "NC_001224.1"))

    rbind(mergeparents) %>% arrange(CHROM, POS) %>%
      select(CHROM, POS, REF, ALT, parent) %>%
      merge(ChromKey) %>% select(-CHROM) %>%
      mutate(CHROM = chromosomes) %>% select(-chromosomes) -> ParentalVCF

  }else{
    rbind(mergeparents) %>% arrange(CHROM, POS) %>%
      select(CHROM, POS, REF, ALT, parent) -> ParentalVCF
  }

  ParentalVCF %>% pivot_wider(names_from = parent, values_from = ALT) -> SNPids

  SNPids$Type <- 0

  for(i in c(P1l, P2l)){
    #filter rows in which all values of columns of the parent NOT selected are NA
    select(SNPids,-i, -CHROM, -POS, -REF) -> tempdf
    tempdf$Any_NA <- apply(tempdf, 1, function(x) anyNA(x))
    SNPids$Type[which(tempdf$Any_NA)] <- i
    rm(tempdf)
  }


  #Collect it to output
  if(Truncate == TRUE){
    SNPids %>% select(CHROM, POS,  Type) %>% filter(Type != 0) -> SNPids
  }

  return(SNPids)

} #*
```

## Load and process raw data

`mydatatotest` is the path/name of the raw output.table file that can be processed by `cybrInputGATKTable()`. 
```{r, message=FALSE, warning=FALSE}
mydatatotest = "TestFiles/XI.SortedCat.vcf.output.table"

cybrInputGATKTable(mydatatotest) %>% mutate(Coverage = as.numeric(AD.REF) + as.numeric(AD.ALT)) %>%
  select(POS, CHROM, Dataset, GQ, AD.REF, AD.ALT, Coverage) -> rawdata

```

Next, convert the parental alleles either with oak and wine (included) or as text files. 

Note: `cybrConvertParentalAlleles_text` **must** be run from within the working directory of the Wine_VCF.txt and Oak_VCF.txt directories, as it uses the names of the files to fill in the allele names downstream. 

```{r, eval = FALSE}
setwd("TestFiles")
parentSNPids <- cybrConvertParentalAlleles_text(Truncate = TRUE)
```
To avoid this issue, use `cybrConvertParentalAlleles_n` with .Rdata files instead:

```{r, message=FALSE, warning=FALSE}
parentSNPids <- cybrConvertParentalAlleles_n(Truncate = TRUE)

head(parentSNPids)
```

```{r, message=FALSE, warning=FALSE}
rawdata %>% 
  merge(parentSNPids) %>% 
  filter(grepl("HNGLVDRXY", Dataset)) %>% #Filter for smaller dataset
  mutate(REFW = as.numeric(Type == "Wine"), REFO = as.numeric(Type == "Oak")) %>%
  group_by(Dataset, CHROM, POS) %>%
  mutate(Wine = max(REFW * as.numeric(AD.REF), REFO * as.numeric(AD.ALT)),
         Oak = max(REFW * as.numeric(AD.ALT), REFO * as.numeric(AD.REF))) %>%
  select(Dataset, POS, CHROM, Coverage, Wine, Oak) %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "Reads") -> rawdata_called

head(rawdata_called)
```

## Smooth data based on gaussian kernel

Sequencing reads are stochastic, and to smooth our data for accuracy, we use a gaussian kernel in `cybr_weightedgass()`. 

```{r, message=FALSE, warning=FALSE}
rawdata_called %>% group_by(Dataset, CHROM, Allele) %>% arrange(POS) %>%
  reframe(POS = POS, 
          SmoothCount = ceiling(frollapply(Reads, n = 200, FUN = cybr_weightedgauss, align = "center"))
          ) -> rawdata_smoothed 

```

Keep in mind that the indices corresponding to the first half of the window size (n) will be NAs
```{r, message=FALSE, warning=FALSE}
head(rawdata_smoothed)
```

With NAs removed:
```{r, message=FALSE, warning=FALSE}
rawdata_smoothed %>% na.omit() %>% head()
```

## Format data for glm

This dataset from _Buzby et al 2024_ is labeled with non-uniform sequencing run names. Here we filter the data for a single run, which corresponds with the specific parents in copper resistance.

```{r, message=FALSE, warning=FALSE}
# Pick out the specific groups of the dataset
rawdata_smoothed %>% 
    mutate(Dataset = gsub("HNGLVDRXY_n01_CuSO4_CSSI_", "", Dataset)) %>%
    mutate(Dataset = gsub(".fastq", "", Dataset)) %>%
    mutate(Dataset = gsub("Unselected", "Unselected_", Dataset)) %>%
    mutate(Dataset = gsub("Selected", "Selected_", Dataset)) %>%
  separate(Dataset, into = c("Bulk", "Parent"), sep = "_") %>% 
  mutate_if(is.character, as.factor) -> rawdata_glm_prep
```

The glm() finds effects of the Bulk (selected or not selected), Parental Background (Oak or Wine), and Interaction terms. Selection is set to 0 or 1 to for ease of interpretation. Because the parents are equal, we set their contrasts to -0.5 and 0.5.

```{r, message=FALSE, warning=FALSE}
contrasts(rawdata_glm_prep$Parent) <- matrix(c(-0.5, 0.5))
contrasts(rawdata_glm_prep$Bulk) <- matrix(c(1,0))
```

## Run logistic regression

Here you can use either `glm_cb2_short()` or `glmer_cb2_short()`, if your model is mixed effect.

```{r, message=FALSE, warning=FALSE}
# Run full dataset
rawdata_glm_prep %>% na.omit() %>% 
  filter(CHROM %in% c("M", "I") == FALSE) %>%
  group_by(CHROM, POS) %>%
  mutate_if(is.character, as.factor) %>%
  summarize(summary = glm_cb2_short(Allele = Allele,
                             Bulk = Bulk,
                             Parent = Parent,
                             #Rep = Rep,
                             W = SmoothCount,
                             formula = "Allele ~ Bulk * Parent",
                             outputlength = 4),
            #MAKE SURE THIS IS THE SAME LENGTH AS OUTPUT LENGTH
            Factor = (c("Intercept", "Bulk", "Parent", "Interaction"))) -> processed_glm_all
```

## Permutations

```{r, message=FALSE, warning=FALSE}
rawdata_glm_prep %>% na.omit() %>% 
  filter(Bulk == "Unselected") %>%
  filter(CHROM %in% c("M", "I") == FALSE) %>%
  group_by(CHROM, POS) %>%
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Bulk, Parent) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Selected") -> rd_shuffled_selected

rawdata_glm_prep %>% na.omit() %>% 
  filter(Bulk == "Unselected") %>%
  filter(CHROM %in% c("M", "I") == FALSE) %>%
  group_by(CHROM, POS) %>%
  pivot_wider(names_from = Allele, values_from = SmoothCount) %>% unnest() %>%
  mutate(label = paste(CHROM, POS, sep = "_")) %>%
  group_by(Bulk, Parent) %>%
  summarize(NewPOS = sample(label),
            Oak = Oak,
            Wine = Wine,
            Parent = Parent) %>%
  separate(NewPOS, into = c("CHROM", "POS"), sep = "_") %>%
  mutate(Bulk = "Unselected") -> rd_shuffled_unselected

rbind(rd_shuffled_selected,rd_shuffled_unselected) %>%
  ungroup() %>%
  pivot_longer(c(Oak, Wine), names_to = "Allele", values_to = "SmoothCount") %>% unnest() %>%
  mutate(POS = as.numeric(POS)) %>%
  mutate_if(is.character, as.factor) -> Perm1contrasts

#Set contrasts for experiment
contrasts(Perm1contrasts$Parent) <- matrix(c(-0.5, 0.5))
contrasts(Perm1contrasts$Bulk) <- matrix(c(1,0))
```

```{r, message=FALSE, warning=FALSE}
processed_glm_all$zscore <- processed_glm_all$summary
processed_glm_all$CSS <- "I"
processed_glm_all$label <- processed_glm_all$Factor

testpeaks <- cybr_lmpeaks(processed_glm_all[processed_glm_all$Factor == "Bulk",], width = 300)

```

# Visualization

```{r, message=FALSE, warning=FALSE, fig.width=8}
processed_glm_all %>% 
  ggplot(aes(x = POS, y = abs(summary), color = Factor)) + geom_line() +
  facet_grid(~CHROM, space = "free", scales = "free") + ggtitle("BSA Bulk Z Scores")
```



```{r, include = FALSE}
cybr_circos2 <- function(d1, d8, peaklist1 = NULL, peaklist8 = NULL, 
                         maxy = NULL, color1 = "#24588A", color8 = "#ED7B01", opacity = "50"){
  
  color1fade_50 <- paste(color1, "50", sep = "")
  color8fade_50 <- paste(color8, "50", sep = "")
  
  color1fade <- paste(color1, opacity, sep = "")
  color8fade <- paste(color8, opacity, sep = "")

  ChromScale <- data.frame(CHROM = factor(as.character(as.roman(1:16)),
                                              levels = as.character(as.roman(1:16))),
                               start = rep(1, 16),
                               end = c(.23,.81,.32, 1.53, 
                                       .58, .27, 1.09, .56, 
                                       .44, .75, .67, 1.08, 
                                       .92, .78, 1.09, .95)*1000000) %>%
  pivot_longer(c(start, end), names_to = "label", values_to = "POS") %>%
    mutate(summary = 0) %>%
    mutate(label = "Interaction") %>%
    transmute(sectors = CHROM,
                 x = POS,
                 y = summary,
                  label = label)
  
  if(is.null(d8)){
    d8 <- d1
    include8 <- FALSE
  }else{
    include8 <- TRUE
  }
    #SET UP DATA
  df8 <- data.frame(sectors = as.character(d8$CHROM),
                 x = d8$POS,
                 y = abs(d8$summary),
                  label = d8$label) %>% rbind(ChromScale)

  df1 <- data.frame(sectors = as.character(d1$CHROM),
                 x = d1$POS,
                 y = abs(d1$summary),
                  label = d1$label) %>% rbind(ChromScale)

  #Take just the interactions of each
  df8int <- df8 %>% filter(label == "Interaction")
  df1int <- df1 %>% filter(label == "Interaction")

    #REORDER THE CHROMOSOMES
  df8int$sectors <- factor(df8int$sectors, levels = as.character(as.roman(1:16)))
  df1int$sectors <- factor(df1int$sectors, levels = as.character(as.roman(1:16)))
  
  ##############################################################################
  dfall <- rbind(df1int, df8int) %>% na.omit()
  circos.par("track.height" = 0.3, start.degree = 90, cell.padding = c(0,0))
  circos.initialize(sectors = dfall$sectors, x = dfall$x)

  if(is.null(maxy)){
    #I think this makes the sizes?
    circos.track(ylim = c(max(c(dfall$y)), 0), dfall$sectors, y = dfall$y, 
        panel.fun = function(x, y) {
            circos.text(CELL_META$xcenter, 
                0 - mm_y(5), 
                CELL_META$sector.index,
                niceFacing = FALSE)
            circos.axis(labels.cex = 0.1)
    })
  }else{
    circos.track(ylim = c(maxy, 0), dfall$sectors, y = dfall$y, 
      panel.fun = function(x, y) {
          circos.text(CELL_META$xcenter, 
              0 - mm_y(5), 
              CELL_META$sector.index,
              niceFacing = FALSE)
          circos.axis(labels.cex = 0.1)
      })
  }
  #Makes the chromosome overlap parts
  #CHROMOSOME I
  draw.sector(83.5, #RIGHT
              90, #LEFT
              rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = color1fade_50, border = FALSE)
  if(include8 == TRUE){
    #CHROMOSOME VIII
    draw.sector(289.5, #LEFT
              305.4, #RIGHT
              rou1 = 0.99, rou2 = 0.69, clock.wise = FALSE, col = color8fade_50, border = FALSE)
  }
  
  
  #Makes the lines
  circos.trackPoints(df8int$sectors, df8int$x, abs(df8int$y), col = color8, pch = 16, cex = 0.1)
  circos.trackPoints(df1int$sectors, df1int$x, abs(df1int$y), col = color1, pch = 16, cex = 0.1)
  
  if(is.null(peaklist8) == FALSE){
    if(length(peaklist8$POS) >= 1){
      for(i in 1:length(peaklist8$POS)){
      circos.link(peaklist8$CHROM[i], 
                  peaklist8$POS[i], 
                  #Add 8 after
                  "VIII", c(0, max(dfall$x[dfall$sectors == "VIII"])),  
                  
                  col = color8fade, 
                  h.ratio = 0.3, 
                  border = color8fade, 
                  lwd = 1)
      }
    }else{
      print("No interactions on Chr 8")
    }
    
  }

  if(is.null(peaklist1) == FALSE){
    if(length(peaklist1$POS) >= 1){
      for(i in 1:length(peaklist1$POS)){
        circos.link("I",c(0, max(dfall$x[dfall$sectors == "I"])),  
                    #add 1 first
                    peaklist1$CHROM[i], 
                    peaklist1$POS[i], 
                    col = color1fade, 
                    h.ratio = 0.3, 
                    border = color1fade, 
                    lwd = 1)
      }
    }else{
      print("No interactions on Chr 1")
    }
  }  
  
  circos.clear()
}
```

```{r, message=FALSE, warning=FALSE}
processed_glm_all$label <- processed_glm_all$Factor
processed_glm_all <- processed_glm_all %>% select(-Factor)

d1 = processed_glm_all
cybr_circos2(d1 = processed_glm_all, 
            d8 = processed_glm_all,
            peaklist8 = testpeaks)
```
